### PR TITLE
User story 12

### LINKED ISSUE #33: User Story 12, Add a New Book to the System
```
As a Visitor,
When I visit the book index page,
I see a link that allows me to add a new book.
When I click that link, I am taken to a new book path.
I can fill in a form to add a new book.
The form includes the following items:
- title
- year book was published
- number of pages (0 or higher)
- image of the book cover (optional, can be left blank)
- a list of authors (see notes below)

When I submit the form, I am taken to that book's show page.

Book titles should be converted to Title Case before saving.
Book titles should be unique within the system.
For authors, a comma-separated list of names should be entered,
and each author will be added to the database.
Authors added to the database should have their names converted
to Title Case.
Author Names should be unique within the system.
If image field is left blank, a default image should be set.
```

DO THESE STORIES FIRST:
- User Story 10, Book Show Page

### Checklist:

- [x] base code tests are written
- [x] base code is written to pass tests
- [x] edge case ("sad path") tests written
- [x] edge case code written
- [x] code is reviewed
- [x] branch is merged
- [x] all tests still pass

child of #18

child of #33

depends on #20

### PR DESCRIPTION
**Pull request by DavisC0801**
This PR/branch...

resolves #33 (Add new book to the system)
Adds a new book view, which is a form for adding new books.
Adds create/new routes and book controller methods
Redirects to new book show page if creation was successful, redirects back to new book form with a message if now.
Tests all of the above.

### CODE REVIEW

File: app/controllers/books_controller.rb [hunk 1]
```diff
@@ -10,4 +10,34 @@ def show
...
+  end
+
+  def create
+    book = Book.new(
+      year_published: book_params[:year_published],
+      page_count: book_params[:page_count],
+      thumbnail: book_params[:thumbnail].presence || 'https://nnp.wustl.edu/img/bookCovers/genericBookCover.jpg',
+      title: book_params[:title].titlecase)
+    if book.save
+      author_input = book_params[:authors].split(",")
```
Comment (chakeresa) [inline app/controllers/books_controller.rb]:
```suggestion
      author_name_inputs = book_params[:authors].split(",")
```

File: app/controllers/books_controller.rb [hunk 2]
```diff
@@ -10,4 +10,34 @@ def show
...
+
+      author_input.each do |author|
```
Comment (chakeresa) [inline app/controllers/books_controller.rb]:
```suggestion
      author_name_inputs.each do |author_name_input|
```

File: app/controllers/books_controller.rb [hunk 3]
```diff
@@ -10,4 +10,34 @@ def show
...
+        book.authors << Author.find_or_create_by(name: author.titlecase.strip)
```
Comment (chakeresa) [inline app/controllers/books_controller.rb]:
```suggestion
        book.authors << Author.find_or_create_by(name: author_name_input.titlecase.strip)
```

File: app/controllers/books_controller.rb [hunk 4]
```diff
@@ -10,4 +10,34 @@ def show
...
+      end
+
+      redirect_to book_path(book)
+    else
+      flash[:notice] = "This book has already been created"
```
Comment (chakeresa) [inline app/controllers/books_controller.rb]: Hm. I'm not sure this code is actually accomplishing what the user story says. I don't see any check that the book title is unique. I think the `else` here is just that the book failed to be created (for any reason), so that's probably what the flash should say. @DavisC0801

File: app/models/book.rb [hunk 5]
```diff
@@ -3,9 +3,10 @@ class Book < ApplicationRecord
   has_many :reviews
   has_many :authors, through: :book_authors
 
-  validates_presence_of :title
-  validates_presence_of :page_count
-  validates_presence_of :year_published
+  validates_presence_of :title, :page_count, :year_published
+  validates_uniqueness_of :title
+  validates :year_published, numericality: { only_integer: true, greater_than_or_equal_to: -4000, less_than_or_equal_to: 2019 }
```
Comment (chakeresa) [inline app/models/book.rb]: Meh, why restrict to <= 2019? Then users in 2020 couldn't add books that were written then. I'd say delete that restriction or do `Time.now.year`.

File: spec/models/book_spec.rb [hunk 6]
```diff
@@ -5,6 +5,9 @@
     it {should validate_presence_of :title}
     it {should validate_presence_of :page_count}
     it {should validate_presence_of :year_published}
+    it {should validate_uniqueness_of :title}
+    it {should validate_numericality_of :year_published}
```
Comment (chakeresa) [inline spec/models/book_spec.rb]: We need to add tests of the other validations you added for both year_published and page_count.

File: app/views/books/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+<h1>New Book</h1>
+<%= form_for @book do |f| %>
+  <% if flash[:notice] %>
+    <p><%= flash[:notice] %></p>
+  <% end %>
+  <%= f.label :title%>
+  <%= f.text_field :title, type: :name, required: true %>
```
Comment (chakeresa) [inline app/views/books/new.html.erb]: What is a `name` type field? I don't see that in the normal HTML input types.

File: app/views/books/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  <br>
+  <%= f.label :year_published %>
+  <%= f.text_field :year_published, type: :number, min: -4000, max: 2019, required: true%>
```
Comment (chakeresa) [inline app/views/books/new.html.erb]: Same issue as the model itself - either don't add a max for the year or do an ERB block to do `Time.now.year`.

File: app/views/books/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  <br>
+  <%= f.label :page_count %>
+  <%= f.text_field :page_count, type: :number, min: 1, required: true %>
+  <br>
+  <%= f.label :thumbnail %>
```
Comment (chakeresa) [inline app/views/books/new.html.erb]: Consider changing the label to say something like "Thumbnail URL", since the user doesn't actually upload an image file.

File: spec/features/books/new_spec.rb [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+      new_book = Book.last
+
+      within "#book-show" do
+        expect(page).to have_link("Robert Fagels")
+        expect(page).to have_link("Homer")
+      end
+    end
+
+    it "converts author names to title case" do
```
Comment (chakeresa) [inline spec/features/books/new_spec.rb:105]: I don't think this test is actually testing what it says it is, haha.

File: spec/features/books/new_spec.rb [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+      expect(new_book.thumbnail).to eq(@thumbnail)
+      expect(new_book.year_published).to eq(@year_published)
+
+      expect(new_book.authors.first.id).to eq(homer.id)
+    end
+
+    it "sets a default image if none is given" do
+      visit new_book_path
+
+      fill_in :book_title, with: "the illiad"
```
Comment (chakeresa) [inline spec/features/books/new_spec.rb:151]: Consider changing to `@title` instead of hard coding

File: spec/features/books/new_spec.rb [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+      fill_in :book_title, with: @title
+      fill_in :book_year_published, with: @year_published
+      fill_in :book_page_count, with: @page_count
+      fill_in :book_thumbnail, with: @thumbnail
+      fill_in :book_authors, with: @authors
+
+      click_button "Create Book"
+
+      new_book = Book.last
+      expect(current_path).to eq(book_path(new_book))
```
Comment (chakeresa) [inline spec/features/books/new_spec.rb:36]: Add RSpec description (or make this a separate `it` statement) that the page redirects to the book show page after successful book creation.

File: app/views/books/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  <%= f.text_field :thumbnail, type: :url %>
+  <br>
+  <%= f.label :authors %>
```
Comment (chakeresa) [inline app/views/books/new.html.erb]: Consider adding a message that prompts the user to enter the authors separated by commas.

File: app/controllers/books_controller.rb (hunk 1 above)
Comment (DavisC0801) [reply app/controllers/books_controller.rb]: Agreed, fixed on branch.

File: app/controllers/books_controller.rb (hunk 4 above)
Comment (DavisC0801) [reply app/controllers/books_controller.rb]: Added new flash edge case for general book creation failure

File: app/models/book.rb (hunk 5 above)
Comment (DavisC0801) [reply app/models/book.rb]: Great point. Corrected on branch.

File: spec/models/book_spec.rb (hunk 6 above)
Comment (DavisC0801) [reply spec/models/book_spec.rb]: Done, added on branch.

File: spec/features/books/new_spec.rb (hunk 10 above)
Comment (DavisC0801) [reply spec/features/books/new_spec.rb:105]: Fixed XD

File: spec/features/books/new_spec.rb (hunk 12 above)
Comment (DavisC0801) [reply spec/features/books/new_spec.rb:36]: Separated into a new test.
